git-credential interfaces (datalad's in Git and vice versa) (#5796)
datalad/datalad: docs/source/credentials.rst

Credentials
***********

Integration with Git
====================

Git and DataLad can use each other's credential system.
In order to allow Git to query credentials from DataLad, the git credential helper delivered with DataLad needs to be configured.
That is, a section like this needs to be part of one's git config file::

>   [credential "http://*.ds.inm7.de"]
>     helper = "datalad"
>   [credential "https://*.ds.inm7.de"]
>     helper = "datalad"

Author: This makes the impression there is always a need to have two config section pairs, but this just for this particular case

```suggestion
  [credential "https://*.data.example.com"]
    helper = "datalad"
```

Note:

- This most likely only makes sense at the user or system level (options `--global`|`--system` with `git config`), since cloning of a repository needs the credentials before there is a local repository.
- The name of that section is a URL matching expression - see `man gitcredentials`.
- The URL matching does NOT include the scheme! Hence, if you need to match `http` as well as `https`, you need two such entries.
- Multiple different git credential helpers can be configured - Git will ask them one after another. For example on OSX, Git comes with a helper to use the OSX keychain which is configured at system level. This does not conflict with setting up DataLad's credential helper.

The other way around, DataLad can ask Git for credentials it knows by other means (other git credential helpers).
For that a datalad provider config needs to be set up in similar fashion::

>   [provider:inm7_ds]
>     url_re = http.*://.*ds\.inm7\.de
>     authentication_type = http_basic_auth
>     credential = inm7_ds
>   [credential:inm7_ds]
>     type = git

Author: ```suggestion
  [provider:data_example_com]
    url_re = http.*://.*data\.example\.com
    authentication_type = http_basic_auth
    credential = data_example_com
  [credential:data_example_com]
    type = git
```

> Such a config lives in a dedicated file named after the provider name (`inm7_ds.cfg` in this case).

Author: ```suggestion
Such a config lives in a dedicated file named after the provider name (`data_example_com.cfg` in this case).
```

> Valid locations for these files are in a dataset (under `.datalad/providers/`), at the user level (under `{user_config_dir}/providers`), at the system level (under `{site_config_dir}/providers`) or come packaged with the datalad distribution (in directory `configs` next to `providers.py`).

Author: I would vote for not duplicating this info here, but to reference the design doc for locations.

Credentials
***********

Integration with Git
====================

> Git and DataLad can use each other's credential system.

Reviewer: It's probably good to mention up from whether either method can be used, or if /both/ are needed and work in tandem.

After reading all of this file, I realized I was unsure about whether both configurations are needed, or if it's one vs the other. I came back to the beginning hoping for a clarification, but didn't find it.

By now, I'm pretty convinced that both are needed, but it feels circular (git says talk to DataLad and DataLad says talk to git). The opening text should make the situation clear.

Reviewer 2: Depends on what you mean by "needed". Both directions work independently and can stand on their own. What's needed depends on what you want. If you want to seemlessly use the credentials provided by both systems, no matter whether that's a `git clone` or a `datalad get`, then yes - you likely want both. But probably not both for the same URL.

In case of our datastores, you areadly provide credentials via a git-credential helper. So, all that's needed to make it complete, is telling DataLad to ask Git (when downloading content) when it encounters such URLs.

Will try to make that more clear.

Reviewer: Ok. I think I see how this works now. The intro could benefit by making one thing explicit: these methods are useful if either DataLad or Git are already configured to discover credentials to access a remote resource. Then the following steps explain how to direct one to the other to get that info.

Obviously that is not exactly the smoothest way of saying it.

I recommend having the intro explain the overall concept. Then one section for each method.

In order to allow Git to query credentials from DataLad, the git credential helper delivered with DataLad needs to be configured.

> That is, a section like this needs to be part of one's git config file::

Reviewer 2: Yes, that's what I mean.

> That is, a section like this needs to be part of one's git config file::

Reviewer: Cool. Then feel free to copy my rewording and use it instead.

> That is, a section like this needs to be part of one's git config file::

Reviewer: Ahh; I /just/ now found the view where GitHub shows me your changes /and/ our comments. It seems like you already lifted the text. LGTM.

  [credential "https://*.data.example.com"]

>     helper = "datalad"

Reviewer 2: > This surprises me. The command datalad is enough to act as a git credential helper? I don't understand how that would work.

See `man gitcredentials` (value string does not begin with an absolute path). It's not a command. It results in calling `git-credential-datalad`. 

> Also, this example assumes that datalad is always in the path, which might not be the case with scripts, cron, etc.

At least the helper (technically not datalad itself) has to be in the path, yes. Otherwise one would need to give the absolute path to that helper. Not good for an example, as it may depend on what system you installed DataLad (and how). So, I wouldn't suggest any absolute path here.

>     helper = "datalad"

Reviewer: > See man gitcredentials (value string does not begin with an absolute path). It's not a command. It results in calling git-credential-datalad.

Ahh yeah, I remember reading that and disliking it. Now I am reminded why. ;-)

Note:

- This most likely only makes sense at the user or system level (options `--global`|`--system` with `git config`), since cloning of a repository needs the credentials before there is a local repository.
- The name of that section is a URL matching expression - see `man gitcredentials`.

> - The URL matching does NOT include the scheme! Hence, if you need to match `http` as well as `https`, you need two such entries.

Reviewer 2: > The necessity of this note is debatable, as it should be covered in man gitcredentials.

I'd argue: Evidently it's sufficiently unclear in that man page, that *you* didn't get it the first time (remember the initial git-credential config you had on juseless?), which is exactly why I put it here. ;-)

> If you /do/ feel it is noteworthy enough, then I propose warning people that they should avoid the http scheme, as credentials will be sent unencrypted (bad™).

I don't think this belongs here. A datalad user does not per se have a choice what URL(s) need(s) to be used. This is about infrastructure, setting up redirects, etc. - not a matter of how to use datalad.

> - Multiple different git credential helpers can be configured - Git will ask them one after another. For example on OSX, Git comes with a helper to use the OSX keychain which is configured at system level. This does not conflict with setting up DataLad's credential helper.

Reviewer 2: > "Git will ask them one after another": does this mean all will always be asked? Or will it query them starting with the first, and attempt them sequentially until one matches and succeeds?

This. Which relates to the suspicion of cycles. Will change wording accordingly.

> OSX -> macOS (they've dropped the OSX for years now, as they are now on version 11.x)

Well. But the helper I am referring to is still called `git-credential-osxkeychain`. Will try to reword whie aowing to identify the one I am talking about.

> "This does not conflict with setting up DataLad's credential helper." I don't quite know what this means (and confuses me more than it helps me). How would the keychain "conflict"? It's just yet another credential helper that can be invoked by DataLad/git, yes?

Yes. I thought it may be a natural question: "There already is a helper configured. Am I supposed to overwrite that config?", because in a lot of cases, a git config you set up, will take precedence over the system config. Not in this case: It's an addition.

The other way around, DataLad can ask Git for credentials it knows by other means (other git credential helpers).
For that a datalad provider config needs to be set up in similar fashion::

  [provider:data_example_provider]

>     url_re = http.*://.*data\.example\.com

Reviewer 2: > What happens if the scheme is https only, but an http connection is initiated and the server redirects to https. Will it match this regex and provide credentials only for the https attempt (as git [correctly] does)?

Not currently. DataLad matches the given URL, not the one it figures in the preocess (when being redirected). But this is outside scope of this PR and a general issue of the cred/downloader system in DataLad.

>     authentication_type = http_basic_auth

Reviewer: Because of this, I /really/ think we should not be providing an example that matches `http`. People tend to copy and paste without groking.

    credential = data_example_cred
  [credential:data_example_cred]

>     type = git

Reviewer: Just to make sure I understand this right, this is where DataLad says "go ask git and see if it has a credential helper for this situation".

If that is the case, I'm pretty convinced that both configurations are needed in order for this to be functional (please correct me if I'm wrong; second time reading through).

It also makes me wonder what kind of weird bugs/corner-cases there could be from having two different regex patterns to match URLs (git's vs Python's). They won't always match exactly the same URLs. Cascading could get very, very weird.

Reviewer 2: I am not sure, where you get the idea from, that both is needed and particularly for the same URL.

However, there probably are cases where a circular setup looks like what one wants. What I can imagine would be a not URL specific config `credential.helper="datalad"` for Git and then encounter a URL for which DataLad is configured to ask Git. In any case: Only makes sense if there's another git-credential-helper that can provide something. That one would need to be configured first, so Git doesn't ask DataLad when it has something else to ask, that only talks to Git.

I will have another look into consequences of circular setups.

Such a config lives in a dedicated file named after the provider name (:file:`data_example_provider.cfg` in this case).
Valid locations for these files are listed in :ref:`chap_design_credentials`.
In opposition to Git's approach, `url_re` is a regular expression that matches the entire URL including the scheme.

> The name of the provider section (`data_example_provider`)needs to match the file name.

Reviewer 2: There's no doc style guide.

The name of the credential section (`data_example_cred`) needs to match the `credential` entry in the provider section.

> DataLad will create a provider configuration interactively, whenever it first encounters an URL that requires authentication and no credentials can be matched yet.

Reviewer 2: > Do you mean that it will prompt the user?

Yes.

> Can that behavior be suppressed?

By providing it beforehand.

This behavior extends to the git credential helper and can therefore even be triggered by a `git clone`.

> However, `git-credential-datalad` will not do this, if the config variable `datalad.credentials.githelper.noninteractive` is set.

Reviewer 2: > Why?

Hm. That's actually a bit tricky as a serious answer (as opposed to "To not be asked!") requires to point out, how this interactivity is different from within such a helper than from within DataLad as a toplevel process (in which case it's auto-determined based on whether we are running in an interactive session).
Don't think that's helpful here. Will just mention that it's intended use is DataLad's CI tests.

Note:

- This most likely only makes sense at the user or system level (options `--global`|`--system` with `git config`), since cloning of a repository needs the credentials before there is a local repository.
- The name of that section is a URL matching expression - see `man gitcredentials`.
- The URL matching does NOT include the scheme! Hence, if you need to match `http` as well as `https`, you need two such entries.
- Multiple git credential helpers can be configured - Git will ask them one after another until it got a username and a password for the URL in question. For example on macOS, Git comes with a helper to use the system's keychain and Git is configured system-wide to query `git-credential-osxkeychain`. This does not conflict with setting up DataLad's credential helper.
- The example configuration requires `git-credential-datalad` to be in the path in order for Git to find it. Alternatively, the value of the `helper` entry needs to be the absolute path of `git-credential-datalad`.
- In order to make Git always consider DataLad as a credential source, one can simply not specify any URL pattern (so it's `[credential]` instead of `[credential "SOME-PATTERN"]`)

> The other way around, DataLad can ask Git for credentials (which it will acquire via other git credential helpers).

Reviewer: I recommend making this its own section with a heading.
